src/main/java/core/basesyntax/service/RegistrationServiceImpl.java

```diff
@@ -5,10 +5,51 @@
 import core.basesyntax.model.User;
 
 public class RegistrationServiceImpl implements RegistrationService {
-    private final StorageDao storageDao = new StorageDaoImpl();
+    private final StorageDao storageDao;
+
+    public RegistrationServiceImpl(StorageDao storageDao) {
+        this.storageDao = storageDao;
+    }
 
     @Override
     public User register(User user) {
-        return null;
+        validateUser(user);
+        storageDao.add(user);
+        return user;
+    }
+
+    private void validateUser(User user) {
+        if (user == null) {
+            throw new RegistrationFailedException("Failed to add new user! \n" +
+                    "User must not be null");
+        }
+        if (user.getLogin() == null) {
+            throw new RegistrationFailedException("Failed to add new user! \n" +
+                    "Login should not be null");
+        }
+        if (user.getPassword() == null) {
+            throw new RegistrationFailedException("Failed to add new user! \n" +
+                    "Password should not be null");
+        }
+        if (user.getAge() == null) {
+            throw new RegistrationFailedException("Failed to add new user! \n" +
+                    "Age should not be null");
+        }
+        if (storageDao.get(user.getLogin()) != null) {
+            throw new RegistrationFailedException("Failed to add new user! \n" +
+                    "User with login " + user.getLogin() + " already exists.");
+        }
+        if (user.getLogin().length() < 6) {
+            throw new RegistrationFailedException("Failed to add new user! \n" +
+                    "Login should have at least 6 characters.");
+        }
+        if (user.getPassword().length() < 6) {
+            throw new RegistrationFailedException("Failed to add new user! \n" +
+                    "Password should have at least 6 characters.");
+        }
+        if (user.getAge() < 18) {
+            throw new RegistrationFailedException("Failed to add new user! \n" +
+                    "User should be at least 18 years old!");
```

Comment:
The checklist advises against adding constants that may change over time, such as the maximum user age. The current hardcoded value '18' should be extracted to a constant field with an informative name to avoid magic numbers and to make future changes easier.